packages/amplify-graphql-relational-transformer/src/has-one/has-one-directive-ddb-references-transformer.ts

```diff
@@ -0,0 +1,47 @@
+import { DDB_DB_TYPE } from '@aws-amplify/graphql-transformer-core';
+import {
+  TransformerContextProvider,
+  TransformerPrepareStepContextProvider,
+  TransformerTransformSchemaStepContextProvider,
+} from '@aws-amplify/graphql-transformer-interfaces';
+import { DataSourceBasedDirectiveTransformer } from '../data-source-based-directive-transformer';
+import { DDBRelationalReferencesResolverGenerator } from '../resolver/ddb-references-generator';
+import { setFieldMappingResolverReference, updateTableForReferencesConnection } from '../resolvers';
+import { HasOneDirectiveConfiguration } from '../types';
+import { ensureReferencesArray, getReferencesNodes, registerHasOneForeignKeyMappings, validateParentReferencesFields } from '../utils';
+
+/**
+ * HasOneDirectiveDDBFieldsTransformer executes transformations based on `@hasOne(references: [String!])` configurations
+ * and surrounding TransformerContextProviders for DynamoDB data sources.
+ *
+ * This should not be used for `@hasOne(fields: [String!])` definitions.
+ */
+export class HasOneDirectiveDDBReferencesTransformer implements DataSourceBasedDirectiveTransformer<HasOneDirectiveConfiguration> {
+  dbType = DDB_DB_TYPE;
+
+  prepare = (context: TransformerPrepareStepContextProvider, config: HasOneDirectiveConfiguration): void => {
+    const modelName = config.object.name.value;
+    setFieldMappingResolverReference(context, config.relatedType?.name?.value, modelName, config.field.name.value, true);
+    registerHasOneForeignKeyMappings({
+      transformParameters: context.transformParameters,
+      resourceHelper: context.resourceHelper,
+      thisTypeName: modelName,
+      thisFieldName: config.field.name.value,
+      relatedType: config.relatedType,
+    });
+  };
+
+  transformSchema = (context: TransformerTransformSchemaStepContextProvider, config: HasOneDirectiveConfiguration): void => {
+    validateParentReferencesFields(config, context as TransformerContextProvider);
+  };
+
+  generateResolvers = (context: TransformerContextProvider, config: HasOneDirectiveConfiguration): void => {
+    updateTableForReferencesConnection(config, context);
+    new DDBRelationalReferencesResolverGenerator().makeHasOneGetItemConnectionWithKeyResolver(config, context);
+  };
+
+  validate = (context: TransformerContextProvider, config: HasOneDirectiveConfiguration): void => {
+    ensureReferencesArray(config);
```

Comment:
This method will throw an error when the `fields` are detected in the config but with the error message `Relational database...`. In this code block, it is related to the `references` usage of a DynamoDB, which is a non-relational database. We should modify/branch the error message to indicate the correct validation rule. (In this case it mostly validates `fields` cannot be used with `references` for Dynamodb)
https://github.com/aws-amplify/amplify-category-api/blob/ed44cd66d504e6d29fe74968d24ae1ce2f665d0a/packages/amplify-graphql-relational-transformer/src/utils.ts#L184

Reply:
Thanks for calling this out. I already have an open item in the feature branch PR to track improving error messages. I've added your a link to your comment there to make sure it's included in that PR.